primap2/_merge.py

```diff
@@ -0,0 +1,331 @@
+    present in both DataArrays. If values from the data to merge are already
+    present in da_start they are treated as equal if the relative difference
+    is below the tolerance threshold. The result will use the values present
+     in da_start.
+
+    The function determines non-unique coordinates and iterates over them to
+    try merging for individual values of the coordinates. Where merge for a value
+    fails because of unequal data, the function recursively dives into the other
+    coordinates until single time-series are left which are then analyzed for each
+    time step and errors are logged as desired.
+
+    The function assumes dataarrays have been checked for identical coordinates
+    etc as it is called by DataArray.pr.merge() and Dataset.pr.merge() (and itself).
+
+
+    Parameters
+    ----------
+    da_start: xr.DataArray
+        data to merge into
+    da_merge: xr.DataArray
+        data to merge into ds_start
+    tolerance: float (optional), default = 0.01
+        The tolerance to use when comparing data
+    error_on_discrepancy (optional), default = True
+        If true throw an exception if false a warning and return values from
+        the calling object in cases of conflict.
+    combine_attrs (optional), default = "drop_conflicts"
+        Governs how to combine conflicting attrs. Is passed on to the xr merge
+        functions.
+
+    Returns
+    -------
+        xr.DataArray: DataArray with data from da_merge merged into the calling object
+
+    """
+
+    entity = da_start.attrs["entity"]
+    if "gwp_context" in da_start.attrs:
+        entity = f"{entity} ({da_start.attrs['gwp_context']})"
+    # try to merge
+    try:
+        da_result = xr.merge(
+            [da_start, da_merge],
+            compat="no_conflicts",
+            join="outer",
+        )
+        # make sure we have a DataArray not a Dataset
+        da_result = da_result[entity]
+        # all done as no errors occurred and thus no duplicates were present
+    except xr.MergeError:
+        # print("Doing a merge by coordinates")
+        # we have conflicting data and try to merge by splitting the DataArray
+        # into pieces along coordinate axes and merge for those to isolate the error
+
+        # find non-unique coords to iterate over
+        coords_start = da_start.coords
+        coords_merge = da_merge.coords
+        coords_to_iterate = []
+        for coord in list(coords_start):
+            vals_start = set(coords_start[coord].values)
+            vals_merge = set(coords_merge[coord].values)
+            all_values = vals_start | vals_merge
+
+            if len(all_values) > 1 and coord != "time":
+                coords_to_iterate.append(coord)
+
+        if len(coords_to_iterate) > 0:
+            coord = coords_to_iterate[0]
+            # determine value sin both dataarrays
+            vals_start = set(coords_start[coord].values)
+            vals_merge = set(coords_merge[coord].values)
```

Comment:
potential performance benefit: re-using the sets from lines 77 and 78 (storing them in coords_to_iterate or so, then re-using them here) Making coord values to sets is potentially quite costly because all values have to be compared to all other values, so it is quadratic in the number of values.

Reply:
I now store the values in a dict and reuse them

Reply:
In general the performance of merge might be suboptimal because of the recursive algorithm. We will see how it works on full datasets.